rustls/src/compress.rs

```diff
@@ -223,6 +229,144 @@ mod feat_brotli {
 
+/// An LRU cache for compressions.
+///
+/// The prospect of being able to reuse a given compression for many connections
+/// means we can afford to spend more time on that compression (by passing
+/// `CompressionLevel::Amortized` to the compressor).
+#[derive(Debug)]
+pub struct CompressionCache {
+    #[cfg(feature = "std")]
+    size: usize,
+
+    // first is LRU, last is MRU
+    #[cfg(feature = "std")]
+    inner: Mutex<Vec<Arc<CompressionCacheEntry>>>,
+}
+
+impl CompressionCache {
+    #[cfg(feature = "std")]
+    /// Create a new, empty `CompressionCache` that can hold up to `size` entries.
+    pub fn new(size: usize) -> Self {
+        Self {
+            size,
+            inner: Mutex::new(Vec::with_capacity(size)),
+        }
+    }
+
+    #[cfg(feature = "std")]
+    pub(crate) fn compression_for(
+        &self,
+        compressor: &dyn CertCompressor,
+        original: &CertificatePayloadTls13,
+    ) -> Result<Arc<CompressionCacheEntry>, CompressionFailed> {
+        if !original.context.0.is_empty() || self.size == 0 {
+            // context is a per-connection quantity, and included in the compressed data.
+            // it is not suitable for inclusion in the cache.
+            return Self::uncached_compression(compressor, original);
+        }
+
+        // cache probe:
+        let encoding = original.get_encoding();
+        let algorithm = compressor.algorithm();
+
+        let mut cache = self
+            .inner
+            .lock()
+            .map_err(|_| CompressionFailed)?;
+        for (i, item) in cache.iter().enumerate() {
+            if item.algorithm == algorithm && item.original == encoding {
+                // this item is now MRU
+                let item = cache.remove(i);
+                cache.push(Arc::clone(&item));
+                return Ok(item);
+            }
+        }
+        drop(cache);
+
+        // do compression:
+        let uncompressed_len = encoding.len() as u32;
+        let compressed = compressor.compress(encoding.clone(), CompressionLevel::Amortized)?;
+        let new_entry = Arc::new(CompressionCacheEntry {
+            algorithm,
+            original: encoding,
+            compressed: CompressedCertificatePayload {
+                alg: algorithm,
+                uncompressed_len,
+                compressed: PayloadU24(Payload::new(compressed)),
+            },
+        });
+
+        // insert into cache
+        let mut cache = self
+            .inner
+            .lock()
+            .map_err(|_| CompressionFailed)?;
+        if cache.len() == self.size {
+            cache.remove(0);
+        }
+        cache.push(Arc::clone(&new_entry));
+        Ok(new_entry)
+    }
+
+    #[cfg(not(feature = "std"))]
+    /// Create a dummy compression cache.
+    ///
+    /// The cache is not supported for no_std builds.
+    pub fn new(_unused: usize) -> Self {
+        Self {}
+    }
+
+    #[cfg(not(feature = "std"))]
+    pub(crate) fn compression_for(
```

Comment:
Instead of having one `CompressionCache` impl with two sets of functions, each gated by the `std` configuration, would it be more natural to have a trait and two impls? One that's a no-op cache for no-std and one that does caching? I think the only shared functionality is `uncached_compression` and I think that could be a free-standing fn both use. WDYT?

Reply:
I had a look at this and agree it's an improvement. To prevent this trait and the types it involves (`CertificatePayloadTls13`, `CompressionCacheEntry`), etc leaking into the public API I think I will have to add some localised disabling of the `private_interfaces` lint. I guess that lint is to prevent private traits from accidentally making unusable bits of public API, but hopefully it's alright here?

Reply:
>  I guess that lint is to prevent private traits from accidentally making unusable bits of public API, but hopefully it's alright here?

Yeah, reading the [proposal that added the lints](https://rust-lang.github.io/rfcs/2145-type-privacy.html#private-in-public-lints) I think your summary is correct, and it doesn't seem applicable in this case.

I would be OK with a localized ignore. @djc WDYT?

Reply:
If it turns into a can of worms the original approach didn't seem outrageous to me.

Reply:
This might be easier if you wrap the fields inside `CompressionCache::Enabled` into a private type?

Reply:
Thanks. Settled on an enum-based formula for this which seems the most tidy.